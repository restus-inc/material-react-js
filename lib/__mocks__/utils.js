/* global jest */
const utils = jest.genMockFromModule('../utils');

const newId = () => 'a0123456';
utils.newId = newId;

export default utils;
export { newId };
