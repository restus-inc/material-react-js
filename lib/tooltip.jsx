/**
 * @license
 * Copyright 2021 Restus Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { useState, useEffect, useRef } from 'react';
import ReactDOM from 'react-dom';
import { MDCTooltip } from '@material/tooltip';
import { useMDCComponentRefProp } from './hooks';

function createTooltip(id) {
  const surfaceElement = document.createElement('div');
  surfaceElement.className = 'mdc-tooltip__surface mdc-tooltip__surface-animation';
  const tooltipElement = document.createElement('div');
  tooltipElement.setAttribute('role', 'tooltip');
  tooltipElement.setAttribute('aria-hidden', 'true');
  tooltipElement.appendChild(surfaceElement);
  tooltipElement.setAttribute('id', id);
  return tooltipElement;
}

/**
 * [MDCTooltip component]{@link https://github.com/material-components/material-components-web/tree/master/packages/mdc-tooltip#readme}
 * implemented by react component. This control supports plain tooltip only.
 * @function Tooltip
 * @param {Object} props
 * @param {string} [props.id] The ID of tooltip element. If not specified, a random string
 * will be set automatically.
 * @param {string} props.text The tooltip text.
 * @param {string} [props.className] The class name that is added to the root element.
 * @param {React.MutableRefObject} [props.mdcTooltipRef] MutableRefObject which bind an
 * MDCRipple instance to.
 * @returns {DetailedReactHTMLElement}
 * @example
 * import { Tooltip } from 'material-react-js';
 *
 * function MyLogoutIcon(props) {
 *   return (
 *     <Tooltip text="Logout">
 *       <button className="material-icons mdc-icon-button" aria-label="logout">logout</button>
 *     </Tooltip>
 *   );
 * }
 * @exports material-react-js
 */
export default function Tooltip(props) {
  const [id] = useState(props.id || `mrj-${Math.floor(Math.random() * 36 ** 8).toString(36).padStart(8, '0')}`);
  const mdcTooltipRef = useRef();

  useEffect(() => {
    const tooltipElement = createTooltip(id);
    document.body.appendChild(tooltipElement);
    mdcTooltipRef.current = new MDCTooltip(tooltipElement);
    return () => {
      mdcTooltipRef.current.destroy();
      document.body.removeChild(tooltipElement);
    };
  }, [id]);

  useEffect(() => {
    if (mdcTooltipRef.current) {
      ReactDOM.render(props.text, mdcTooltipRef.current.root.children[0]);
    }
  }, [props.text]);

  useEffect(() => {
    if (mdcTooltipRef.current) {
      const className = props.className ? `mdc-tooltip ${props.className}` : 'mdc-tooltip';
      mdcTooltipRef.current.root.className = className;
    }
  }, [props.className]);

  useMDCComponentRefProp(mdcTooltipRef, props.mdcTooltipRef);

  if (Array.isArray(props.children) || typeof props.children === 'string' || !props.children) {
    return <div aria-describedby={id}>{props.children}</div>;
  }
  return React.cloneElement(props.children, { 'aria-describedby': id });
}
