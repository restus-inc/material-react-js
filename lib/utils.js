/**
 * Utilities for Material React.
 * @module material-react/lib/utils
 */

/**
 * Generates a rundom ID. The generated ID starts with basic latin child alphabet, and
 * subsequent characters consist of child alphanumeric character. (i.e. it matches
 * `/[a-z][a-z0-9]{digit-1}/`)
 * @param {number} [digit=8] The number of digit of the generated ID
 * @returns {string} The ID generated
 */
export function newId(digit = 8) {
  return (
    (Math.floor(Math.random() * 26) + 10) * (36 ** (digit - 1))
    + Math.floor(Math.random() * (36 ** (digit - 1)))
  ).toString(36);
}

/**
 * Converts the event name to event handler name.
 * @param {string} eventName The event name that is converted to the handler name
 * @return {string} The event handler name converted
 */
export function handlerName(eventName) {
  return `on${eventName.charAt(0).toUpperCase()}${eventName.slice(1)}`;
}
