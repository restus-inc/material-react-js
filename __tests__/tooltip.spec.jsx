/**
 * @license
 * Copyright 2020 Restus Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/* global jest describe, beforeEach, afterEach, it, expect */
import 'regenerator-runtime/runtime';
import React, { useEffect, useRef } from 'react';
import { render, cleanup } from '@testing-library/react';
import { MDCTooltip } from '@material/tooltip';
import { htmlOfDocumentBody } from './utils';

import Tooltip from '../lib/tooltip';
import Button from '../lib/button';

describe('Tooltip component', () => {
  beforeEach(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(0.012345);
  });

  afterEach(async () => {
    jest.spyOn(global.Math, 'random').mockRestore();
    await cleanup();
  });

  it('supports plain tooltip for a single React element', () => {
    expect(htmlOfDocumentBody(
      <Tooltip text="foo" id="baz">
        <button>button</button>
      </Tooltip>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfDocumentBody(
      <Tooltip text={<>foo<br/>bar</>} className="qux">
        <Button label="button"/>
      </Tooltip>,
    )).resolves.toMatchSnapshot();
  });

  it('supports plain tooltip for multiple elements', () => {
    expect(htmlOfDocumentBody(
      <Tooltip text="foo" id="baz">
        <button>button1</button>
        <button>button2</button>
      </Tooltip>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfDocumentBody(
      <Tooltip text={<>foo<br/>bar</>} className="qux">
        text element
        <button>button2</button>
      </Tooltip>,
    )).resolves.toMatchSnapshot();
  });

  it('supports plain tooltip for a text elementa', () => {
    expect(htmlOfDocumentBody(
      <Tooltip text="foo" id="baz">
        text element
      </Tooltip>,
    )).resolves.toMatchSnapshot();

    expect(htmlOfDocumentBody(
      <Tooltip text={<>foo<br/>bar</>} className="qux">
        text element
      </Tooltip>,
    )).resolves.toMatchSnapshot();
  });

  it('can provide an MDCTooltip instance', async () => {
    let mdcTooltipComponent;
    function MyTooltip() {
      const mdcTooltipRef = useRef();
      useEffect(() => {
        mdcTooltipComponent = mdcTooltipRef.current;
      });
      return <Tooltip text="foo" mdcTooltipRef={mdcTooltipRef}><button>button</button></Tooltip>;
    }
    render(<MyTooltip/>);
    expect(mdcTooltipComponent).toBeInstanceOf(MDCTooltip);
  });
});
