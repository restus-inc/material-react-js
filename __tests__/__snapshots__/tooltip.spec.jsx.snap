// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component supports plain tooltip for a single React element 1`] = `
"<body>
  <div>
    <button
      aria-describedby=\\"baz\\"
    >
      button
    </button>
  </div>
  <div
    aria-hidden=\\"true\\"
    class=\\"mdc-tooltip\\"
    id=\\"baz\\"
    role=\\"tooltip\\"
  >
    <div
      class=\\"mdc-tooltip__surface mdc-tooltip__surface-animation\\"
    >
      foo
    </div>
  </div>
</body>"
`;

exports[`Tooltip component supports plain tooltip for a single React element 2`] = `
"<body>
  <div>
    <button
      aria-describedby=\\"mrj-0fzyuxxr\\"
      class=\\"mdc-button\\"
    >
      <div
        class=\\"mdc-button__ripple\\"
      />
      <span
        class=\\"mdc-button__label\\"
      >
        button
      </span>
    </button>
  </div>
  <div
    aria-hidden=\\"true\\"
    class=\\"mdc-tooltip qux\\"
    id=\\"mrj-0fzyuxxr\\"
    role=\\"tooltip\\"
  >
    <div
      class=\\"mdc-tooltip__surface mdc-tooltip__surface-animation\\"
    >
      foo
      <br />
      bar
    </div>
  </div>
</body>"
`;

exports[`Tooltip component supports plain tooltip for a text elementa 1`] = `
"<body>
  <div>
    <div
      aria-describedby=\\"baz\\"
    >
      text element
    </div>
  </div>
  <div
    aria-hidden=\\"true\\"
    class=\\"mdc-tooltip\\"
    id=\\"baz\\"
    role=\\"tooltip\\"
  >
    <div
      class=\\"mdc-tooltip__surface mdc-tooltip__surface-animation\\"
    >
      foo
    </div>
  </div>
</body>"
`;

exports[`Tooltip component supports plain tooltip for a text elementa 2`] = `
"<body>
  <div>
    <div
      aria-describedby=\\"mrj-0fzyuxxr\\"
    >
      text element
    </div>
  </div>
  <div
    aria-hidden=\\"true\\"
    class=\\"mdc-tooltip qux\\"
    id=\\"mrj-0fzyuxxr\\"
    role=\\"tooltip\\"
  >
    <div
      class=\\"mdc-tooltip__surface mdc-tooltip__surface-animation\\"
    >
      foo
      <br />
      bar
    </div>
  </div>
</body>"
`;

exports[`Tooltip component supports plain tooltip for multiple elements 1`] = `
"<body>
  <div>
    <div
      aria-describedby=\\"baz\\"
    >
      <button>
        button1
      </button>
      <button>
        button2
      </button>
    </div>
  </div>
  <div
    aria-hidden=\\"true\\"
    class=\\"mdc-tooltip\\"
    id=\\"baz\\"
    role=\\"tooltip\\"
  >
    <div
      class=\\"mdc-tooltip__surface mdc-tooltip__surface-animation\\"
    >
      foo
    </div>
  </div>
</body>"
`;

exports[`Tooltip component supports plain tooltip for multiple elements 2`] = `
"<body>
  <div>
    <div
      aria-describedby=\\"mrj-0fzyuxxr\\"
    >
      text element
      <button>
        button2
      </button>
    </div>
  </div>
  <div
    aria-hidden=\\"true\\"
    class=\\"mdc-tooltip qux\\"
    id=\\"mrj-0fzyuxxr\\"
    role=\\"tooltip\\"
  >
    <div
      class=\\"mdc-tooltip__surface mdc-tooltip__surface-animation\\"
    >
      foo
      <br />
      bar
    </div>
  </div>
</body>"
`;
